src/metabase/query_processor/middleware/results_metadata.clj

```diff
@@ -19,13 +20,24 @@
 ;;; |                                                   Middleware                                                   |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-;; TODO -
-;;
-;; 1. Is there some way we could avoid doing this every single time a Card is ran? Perhaps by passing the current Card
-;;    metadata as part of the query context so we can compare for changes
-;;
-;; 2. Consider whether the actual save operation should be async as with
-;;    [[metabase.query-processor.middleware.process-userland-query]]
+(defn- standardize-metadata
+  "There is sometimes a difference between stored metadata and the 'new' metadata,
+  where the 'new' metadata has nil values for some keys whereas the stored metadata does not include the keys.
+  This function removes those nil valued-keys to avoid false negatives."
+  [metadata]
+  (let [drop-nil-keys #{:coercion_strategy :settings :fk_target_field_id :semantic_type}]
+    (cond
+      (map? metadata)
+      (m/filter-kv (fn [k v] (or (some? v)
+                                 (not (drop-nil-keys k))))
+                   metadata)
+
+      (sequential? metadata)
+      (mapv standardize-metadata metadata)
+
+      :else
+      metadata)))
```

Comment:
This function is still being way more flexible about shapes than it needs to be.

You can just `map` the `m/filter-kv` over each column:

```suggestion
    (for [col metadata]
      (m/filter-kv (fn [k v] (or (some? v)
                                 (not (drop-nil-keys k))))
                    col))))
```